realm/realm-library/src/main/java/io/realm/RealmConfiguration.java

```diff
@@ -594,11 +610,46 @@ public Builder schemaVersion(long schemaVersion) {
             return this;
         }
 
+        /**
+         * Update the schema version of the Realm and apply the provided migration.
+         * <p>
+         * The schema version must be equal to or higher than the schema version of the existing
+         * Realm file, if any, and the supplied {@link RealmMigration} must handled any schema changes.
+         * For <i>version only</i> updates the supplied
+         * {@link RealmMigration#migrate(DynamicRealm, long, long)} can be a <i>no operation</i>.
+         * <p>
+         * If the schema has changed without increasing the version or failing to resolve schema
+         * migration through the supplied {@link RealmMigration},, Realm will throw a
+         * {@link io.realm.exceptions.RealmMigrationNeededException}.
+         *
+         * @throws IllegalArgumentException If {@code schemaVersion} is negative or {@code migration} is null.
+         *
+         * @see #schemaVersion(long)
+         * @see #migration(RealmMigration)
+         */
+        public Builder schemaVersion(long schemaVersion, RealmMigration migration) {
```

Comment:
Did you consider alternatives to this name? Would just `schema(long schemaVersion, RealmMigration migration)` be more accurate? I'm not sure. Maybe it makes sense making "schemaVersion" the primary concept and the migration just a result of that.

Reply:
Also. Since we are in the process of making breaking changes. Would it make sense to prepare the API for https://github.com/realm/realm-java/issues/2530 ?

Should we already now rename `RealmMigration` to perhaps `RealmManualMigration` so in the future we can add `ReamAutomaticMigration` (just throwing ideas at the wall, we could also postpone and decide later. E.g. we can introduce both of these interfaces in a new update and just let RealmMigration be an alias for RealmManualMigration until the next major release).

Reply:
Since the migration would never be fully automated for all cases, I guess a `RealmAutomaticMigration` would also need some customization points, so maybe better off separating the automatism into a helper and allowing people to take advantage of this in other scenarios. Thus, keeping the `RealmMigration` interface for customization but supplying a `DefaultRealmMigration` implementation for simple cases where the automatic migration helper is assumed to handle everything (simple additions/deletions, etc).

Reply:
I think I need to see some code to understand that 😄

Reply:
> Did you consider alternatives to this name? Would just `schema(long schemaVersion, RealmMigration migration)` be more accurate? I'm not sure. Maybe it makes sense making "schemaVersion" the primary concept and the migration just a result of that.

Shorter names are good, but maybe some keeping it to align with other platforms. If not bound by that I would go for `version` but maybe that is too overloaded?

Reply:
JS has separate: `schema`, `schemaVersion` and `migration` blocks
Cocao has: `schemaVersion` and `migrationBlock`
.NET has `SchemaVersion` and `MigrationCallback`

That probably hints to us to keep `schemaVersion(version, migration)`. Thoughts?

Reply:
Yep. Lets just keep it, to avoid too may different names.